mne_connectivity/var.py

```diff
@@ -0,0 +1,406 @@
+
+    This will construct a matrix along the 0th axis (rows),
+    stacking copies of the data based on order.
+
+    Parameters
+    ----------
+    snapshots : np.ndarray
+        A multivariate time-series ``(n_signals, n_times)``.
+    order : int
+        The order of the linear model to be estimated.
+    n_times : int
+        The number of times in the original dataset.
+
+    Returns
+    -------
+    snaps : np.ndarray
+        A snapshot matrix with copies of the original ``snapshots``
+        along the rows based on the ``order`` of the model.
+
+    Notes
+    -----
+    Say ``snapshots`` is an array with shape (N, T) with
+    order ``M``. We will abbreviate this matrix and call it ``X``.
+    ``X_ij`` is the ith signal at time point j.
+
+    The resulting ``snaps`` matrix would be a (N*M, T - M) array:
+
+    +------+------+------+------------+
+    | X_00 | X_01 | ...  | X_0(T-M+1) |
+    +------+------+------+------------+
+    | X_10 | X_11 | ...  | X_1(T-M+1) |
+    +------+------+------+------------+
+    | ...  | ...  | ...  | ...        |
+    +------+------+------+------------+
+    | X_N0 | X_N1 | ...  | X_N(T-M+1) |
+    +------+------+------+------------+
+    | X_01 | X_02 | ...  | X_0(T-M+2) |
+    +------+------+------+------------+
+    | ...  | ...  | ...  | ...        |
+    +------+------+------+------------+
+    | X_N1 | X_N2 | ...  | X_N(T-M+2) |
+    +------+------+------+------------+
+    | ...  | ...  | ...  | ...        |
+    +------+------+------+------------+
+    | X_NM | X_N2 | ...  | X_N(T-M+M-1)|
+    +------+------+------+------------+
+
+    """
+    snaps = np.concatenate(
+        [snapshots[:, i: n_times - order + i + 1] for i in range(order)],
+        axis=0,
+    )
+    return snaps
+
+
+def _system_identification(data, times, names=None, model_order=1, delta=0,
+                           random_state=None, memmap=False, n_jobs=-1,
+                           compute_fb_operator=False,
+                           verbose=True):
+    """Solve system identification using least-squares over all epochs.
+
+    Treats each epoch as a different window of time to estimate the model:
+
+    .. math::
+        X(t+1) = \sum_{i=0}^{order} A_i X(t - i)
+
+    where ``data`` comprises of ``(n_signals, n_times)`` and ``X(t)`` are
+    the data snapshots.
+    """
+    # 1. determine shape of the window of data
+    n_epochs, n_nodes, n_times = data.shape
+
+    model_params = {
+        'delta': delta,
+        'model_order': model_order,
+        'random_state': random_state,
+        'compute_fb_operator': compute_fb_operator
+    }
+
+    # compute the A matrix for all Epochs
+    A_mats = np.zeros((n_epochs, n_nodes * model_order, n_nodes))
+    if n_jobs == 1:
+        for idx in tqdm(range(n_epochs)):
+            A = _compute_lds_func(data[idx, ...], **model_params)
+            A_mats[idx, ...] = A
+    else:
+        try:
+            from joblib import Parallel, delayed, dump, load
```

Comment:
why do you need to go so low level with memmap? The purpose of joblib is to hide this complexity from algorithmic code.

Reply:
Removed memmap.